### PR TITLE
fix: fit tokenizer only on training data to prevent data leakage

### LINKED ISSUE #3035: Improve:
### Is there an existing issue for this?

- [x] I have searched the existing issues

### What would you like to improve?

# Prevent Tokenizer Data Leakage During Scorer Training

## Describe the feature

The scorer training pipeline currently fits the tokenizer on the entire dataset before any train/validation split is performed.

Current implementation:

```python
tokenizer.fit_on_texts(stories + prompts)
```

Since the tokenizer is built using both training and validation samples, information from the validation set is indirectly exposed during training preparation.

## Why this matters

Although the model weights are not trained on validation labels, the vocabulary is still derived from the complete dataset.

This can lead to:

* Data leakage between training and validation sets.
* Overly optimistic evaluation metrics.
* Less realistic measurement of generalization performance.
* Difficulty comparing future experiments using proper ML evaluation practices.

## Proposed Solution

1. Split the dataset into training and validation subsets first.
2. Fit the tokenizer only on the training stories and prompts.
3. Use the fitted tokenizer to transform both training and validation data.
4. Keep validation data completely unseen during vocabulary construction.

Example workflow:

```python
# Split first
X_train, X_val = ...

# Fit tokenizer only on training data
tokenizer.fit_on_texts(train_stories + train_prompts)

# Transform both sets using the same tokenizer
train_encoded = ...
val_encoded = ...
```

## Benefits

* Eliminates vocabulary-level data leakage.
* Produces more reliable validation metrics.
* Aligns the training pipeline with standard machine learning practices.
* Improves experiment reproducibility and model evaluation quality.

## Additional Context

This change does not modify the model architecture or scoring logic. It only improves the correctness of the training and evaluation pipeline.

### Why does this matter?

_No response_

### Screenshots or screen recording (if applicable)

_No response_

### Record

- [x] I agree to follow this project's Code of Conduct
- [x] I want to work on this issue

### PR DESCRIPTION
**Pull request by Yesha-19**
Closes #3035 
## 🐛 Problem

The scorer training pipeline was fitting the tokenizer on the **entire
dataset** before any train/validation split was performed:

```python
tokenizer.fit_on_texts(stories + prompts)  # all 2000 samples
# ...
model.fit(..., validation_split=0.15)       # split happens too late
```

This caused **vocabulary-level data leakage** — words from validation
samples were included in the tokenizer's vocabulary during training
preparation, making the model's evaluation metrics unrealistically
optimistic.

---

## ✅ Fix

Reordered the pipeline to follow correct ML practices:

1. Generate the full dataset
2. **Split into train/val first** using `train_test_split`
3. **Fit tokenizer only on training data**
4. Encode both splits using the same fitted tokenizer
5. Pass explicit `validation_data` to `model.fit()` instead of `validation_split`

```python
# Split first
(train_stories, val_stories, ...) = train_test_split(
    stories, prompts, coh, cre, rel,
    test_size=0.15,
    random_state=42,
)

# Fit only on training data
tokenizer.fit_on_texts(train_stories + train_prompts)

# Encode both sets
train_story_enc = encode(train_stories)
val_story_enc   = encode(val_stories)

# Explicit validation data
model.fit(..., validation_data=(...))
```

---

## 📂 Files Changed

| File | Change |
|---|---|
| `backend/ml/train_scorer.py` | Reordered split → tokenizer fit → encode; replaced `validation_split` with explicit `validation_data`; fixed `model.predict()` to use encoded val arrays; moved `matplotlib` import to top; removed duplicate `numpy` import |

---

## 🔍 Additional Notes

- No changes to model architecture (`scorer.py`) or inference path (`score_api.py`)
- `random_state=42` ensures the split is reproducible across runs
- Val loss after this fix may appear slightly higher than before — this reflects **more honest evaluation**, not degraded performance
- Validation samples that contain words unseen during training will now correctly map to `<OOV>` as expected

---

## ☑️ Checklist

- [x] I have read the [CONTRIBUTING.md](../CONTRIBUTING.md) guidelines
- [x] My branch is up to date with `main`
- [x] The fix addresses the root cause, not just the symptom
- [x] No model architecture or inference code was modified
- [x] Code follows the existing style and conventions of the project
- [x] Imports are clean — no duplicates, all at the top of the file
- [x] The fix is limited in scope to the reported issue only

### CODE REVIEW

**Comment by Yesha-19** [comment]
Hi @ronisarkarexe,

I have submitted a fix for the data leakage issue identified in issue #3035. I would appreciate it if you could review these changes when you have a moment. Please let me know if any further adjustments are required.

Thank you!
